### PR TITLE
Add DarkPoolRatio metrics

### PR DESCRIPTION
**Pull request by marcindsobczak**

### CODE REVIEW

**Review by MarekM25 — APPROVED** [review]

File: src/Nethermind/Nethermind.Blockchain/OnChainTxWatcher.cs [hunk 1]
```diff
@@ -56,12 +57,20 @@ private void OnBlockAddedToMain(object sender, BlockReplacementEventArgs e)
...
+
+            for (int i = 0; i < transactionsInBlock; i++)
             {
-                _txPool.RemoveTransaction(block.Transactions[i].Hash, true);
+                Keccak txHash = block.Transactions[i].Hash;
+                if (!_txPool.IsTransactionKnown(txHash))
+                {
+                    discoveredTransactions++;
+                } 
+                _txPool.RemoveTransaction(txHash, true);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/OnChainTxWatcher.cs:70]: Please modify RemoveTransaction to return bool if transaction was known

File: src/Nethermind/Nethermind.TxPool/Metrics.cs [hunk 2]
```diff
@@ -31,5 +31,11 @@ public static class Metrics
         
         [Description("Number of known pending transactions.")]
         public static long PendingTransactionsKnown { get; set; }
+        
+        [Description("Ratio of discovered transactions in the block")]
+        public static float DarkPoolRatio { get; set; }
+        
+        [Description("Ratio of previously known transactions in the block")]
+        public static float OpenPoolRatio { get { return 1 - DarkPoolRatio; } }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/Metrics.cs:39]: We can calculate OpenPoolRatio on grafana side, we can drop it here.

File: src/Nethermind/Nethermind.TxPool/TxPool.cs [hunk 3]
```diff
@@ -461,6 +461,11 @@ public bool TryGetPendingTransaction(Keccak hash, out Transaction transaction)
             return transaction != null;
         }
 
+        public bool IsTransactionKnown(Keccak hash)
+        {
+            return _hashCache.Get(hash);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/TxPool.cs:466]: Now good question. Is know would be hashCashe? Or maybe we should only check actual held transactions? Or maybe we should also check DB?

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.TxPool/TxPool.cs (hunk 3 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs]: hash cache is better because we have heard about the txs from the hash cache

Comment (tkstanczak) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs]: we can do two-level ratio though

Comment (marcindsobczak) [reply src/Nethermind/Nethermind.TxPool/TxPool.cs]: added checking actual held transactions and needed to move whole logic from OnChainTxWatcher to BlockProcessor (in OnChainTxWatcher an attempt to remove transactions always returns false - they are already removed)

**Review by LukaszRozmej — APPROVED** [review]
Can you post current metrics on mainnet?

**Comment by marcindsobczak** [comment]
Ratios from last 15min:
poolratios

Level1 - HashCache
Level2 - Pending

**Comment by LukaszRozmej** [comment]
> Ratios from last 15min:
> poolratios
> 
> Level1 - HashCache
> Level2 - Pending

1. Lets put them on same graph
2. Change Y axis to %
3. How L1 can be bigger than L2?

**Comment by marcindsobczak** [comment]
Last 1h on one graph with Y in %. Added average and current values too:
poolratiosonegraph
3. It never is. It was just an illusion because of the different Y scales

**Pull request merged**